src/sklearn_knn/original/_gnn.py

```diff
@@ -1,15 +1,12 @@
-from sklearn.pipeline import Pipeline
-
-from .._base import IDNeighborsClassifier, KnnPipeline
+from .._base import IDNeighborsClassifier
 from ._cca_transformer import CCATransformer
 
 
-class GNN(KnnPipeline):
-    def _get_pipeline(self):
-        transformer = CCATransformer()
-        classifier = IDNeighborsClassifier(
-            n_neighbors=self.n_neighbors,
-            weights=self.weights,
-        )
-        steps = [("transform", transformer), ("classifier", classifier)]
-        return Pipeline(steps)
+class GNN(IDNeighborsClassifier):
+    def fit(self, X, y, spp=None):
+        X = CCATransformer().fit_transform(X, y=y, spp=spp)
+        return super().fit(X, y)
+    
+    def predict(self, X, spp=None):
+        X = CCATransformer().fit_transform(X, spp=spp)
+        return super().predict(X)
```

Comment:
@grovduck I think I need to make a few changes here but want to double check. Should `fit` and `predict` both take an optional `y` argument? I'm still a little shaky on the mechanics of the modeling vs mapping matrices, but if I understood right, GNN will always use one or the other to run the ordination, so we need the option for both?

Reply:
I think I missed this in my original review, but it looks like in both `Euclidean` and `GNN`, you're calling `fit_transform` in `predict` rather than just `transform`.  I think that will cause issues if we pass new data because the transformer will get refit, correct?  (I may have had this in the original code).  I think we just want:
```
def predict(self, X):
    X = CCATransformer().transform(X)
    return super().predict(X)
```
And, now that I'm thinking of it, I think this is one of those times we might want to save the transform as part of the call to `fit`, even though it seems really wrong to set `self` attributes outside of `__init__`.  But they do this all over the place in sklearn, such that they have a convention that you give the attribute a trailing underscore if set outside of `__init__` (at least that's what I think they do).  So I think we can even do this:
```
def fit(self, X, y, spp=None):
    self.transform_ = CCATransformer().fit(X, y=y, spp=spp)
    X = self.transform_.transform(X)
    return super().fit(X, y)

def predict(self, X):
    # check for self.transform_
    X = self.transform_(X)
    return super().predict(X)
    # or return super().predict(self.transform_(X))
```
I've tested and it does work.  Let me know if I'm thinking about this incorrectly.

Reply:
For now, I haven't made this change, nor moved stuff out of `original` pending your thoughts.

Reply:
You're spot on, I was thinking in terms of fitting the model rather than fitting the model *and* the transformer. Adding the `transform_` attribute during fitting seems a little wrong to me too, but I agree that's the approach to take. I think adding a [`check_is_fitted(self)`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html) to `predict` will make sure `self.transform_` exists. Also, if you take a look at that function, it refers to the trailing-underscore attrs as "fitted attributes", so I think we're consistent with their convention.

Feel free to go ahead with the change, thanks!

Reply:
Just realized we're not actually calling `predict` yet on these estimators and right now we really need to do `check_is_fitted(self)` on `kneighbors_ids` and `kneighbors` methods.  We could do this:
```
class IDNeighborsClassifier(KNeighborsClassifier):
    def kneighbor_ids(self, X=None, n_neighbors=None):
        check_is_fitted(self, "transform_")
        neigh_ind = super().kneighbors(X, n_neighbors, False)
        return self.classes_[self._y[neigh_ind]]
```
but `Raw` doesn't have a `transform_` attribute, so that would fail.  Otherwise, we'd need to do this on each:
```
class GNN(IDNeighborsClassifier):
    ...
    def kneighbors_id(self, X=None, n_neighbors=None):
        check_is_fitted(self, "transform_")
        return super().kneighbors_id(X,=X, n_neighbors=n_neighbors()
```
This is where I usually tie myself in knots trying to figure out which path to take.  Do you have an opinion?

The other thing I realized is that our tests are only passing right now because we are passing the full dataset on `fit` and they are getting transformed as part of that process.  When we call `kneighbors` with no argument, it is running on that trained data.  However, I think if we pass new data, that transform of the X data won't happen because the call to `kneighbors` is just using `KNeighborsClassifier`.  So I think we need to first get some tests in place to split the data, fit with some and run `kneighbors` on the other to make sure this will work.  I can take a first cut, but it will likely be ugly and require your refactoring.

Reply:
Is there a reason we can't use `check_is_fitted` without specifying any attributes? That would check that the underlying `KNeighborsClassifier` is fitted, which would work for `Raw`. Also, I believe we can avoid the `check_is_fitted` within `kneighbor_ids` because it happens already within `KNeighborsClassifier.kneighbors`.

With that said, I think you're right that calling `kneighbors` with new data will fail because the transformation won't be applied. I'm not quite sure the best way to handle that, but my first thought is something like below that gets added to all the estimators except for `Raw`. There may be an issue with method resolution order when `kneighbor_ids` gets called, but I haven't gotten that far yet... There also may be a simpler, cleaner solution I haven't thought of.

```python
class TransformedKNeighborsMixin(KNeigborsClassifier):
    def kneighbors(X, *args, **kwargs):
        check_is_fitted(self)
        X = self.transform_(X)
        super().kneighbors(X, *args, **kwargs)
```

In any case, I think you're right that the next step should be more comprehensive tests so we at least know what's working and what isn't. I don't think we need to fully polish the dataset and test system necessarily, just as long as we can get them running and sort out the estimators.

Reply:
My read of `check_is_fitted` was that if you don't pass any attributes, it will look for the presence of *any* fitted attribute.  I think any estimator that derives from `KNearestNeighbors` will have (at least) `classes_` and  `outputs_2d_` set as part of its fit, so it would pass just based on this without needed to check for the transform.

The idea of the `TransformedKNeighborsMixin` is where my mind went as well, but I was putting it into a strict class hierarchy rather than a mixin because I'm not as familiar with them.  I like this approach if you think it will work - I think I grok the concern about MRO as well.

I did create some subsetted test data for working with `Raw` which we knew would work and it does.

Reply:
> it would pass just based on this without needed to check for the transform.

This is my understanding too. My thinking is that `check_is_fitted` is a runtime validation to make sure the user fit the model before attempting to get `kneighbors` or run `predict`. As long as we implement things correctly, the `transform_` will always be set when the estimator is fit. Is there an edge case I'm not considering where the estimator could end up getting fit without the `transform_` getting set?

> The idea of the TransformedKNeighborsMixin is where my mind went as well

I'm pretty new to mixins, but it seems like they're widely used in `sklearn` which is mostly why I've been leaning towards them. Aside from that, the only benefit I see is that it's a little clearer that they're never meant to be instantiated directly than if they were just a typical superclass. I can do some experimenting with this to see if this will work in practice and let you know if I hit any roadblocks.

Glad to hear the `Raw` is working as expected!

Reply:
The `TransformedKNeighborsMixin` is now added and seems to be working just fine, and ensures that `kneighbors` runs transforms as expected. Calling this one resolved!